### DISCUSSION TITLE
Support file extension: `.q.md`, `.qmd.md`, or `._q_.md`?

### DISCUSSION THREAD

**Discussion by dipterix** [quarto-dev/quarto-cli#10854] [Feature Requests]
### Description

For this feature, Github would render `md` and `qmd` pretty differently. Would it be nice so people can edit quarto directly on Github?

<img width="142" alt="image" src="https://github.com/user-attachments/assets/0c3e99d1-1956-4bd4-a347-075583708da4">

Our editors were (mostly) born before computer technology becomes so popular. They are not used to Github pull, commit, push, but are very happy to edit directly on Github. With the visual markdown renderers, they can easily see the results. However, `*.qmd` is not yet supported by Github.

Would it possible to support compiling files like `.q.md` like `.qmd`?

**Comment by mcanouil** [comment]
I don't understand. "qmd" files are recognised by GitHub as Markdown-based files, why do you want weird file extensions?
Changing the extensions makes no sense as it won't change anything about Quarto/Pandoc spcific markdown syntax.

- https://github.com/github-linguist/linguist/blob/865ad2ae6e6d5c144e2e2d384f5d68ae211692a0/lib/linguist/languages.yml#L6130-L6141

**Comment by dipterix** [comment]
@mcanouil This is what Github looks like when you edit qmd files. Github treats qmd as code and only shows document diffs. You can't see the render output when editing the files

<img width="1071" alt="image" src="https://github.com/user-attachments/assets/bbf418d5-8f29-4976-81b6-2d10ad6fe5ec">

I don't think "qmd" files are recognised by GitHub as Markdown-based files, otherwise Github will render `qmd`'s markdown components when previewing the file.

This is what you have when editing md files. Github shows the preview when you edit it.

<img width="1071" alt="image" src="https://github.com/user-attachments/assets/9277876f-e978-424b-a689-9984a4f629a9">

My proposal is to let Github edit work naturally with quarto documents.

This is the same idea behind quarto visual editor: you see what you get. However, for non-programmers with zero knowledge on markdowns and programmings, it's quite tricky to get them install quarto properly. Supporting extensions like `.q.md` seems to be an low hanging fruit.

I see in https://github.com/jjallaire/linguist/blob/e689491c945e1cf3b90fabe8c9fdc2d6a41c2769/lib/linguist/languages.yml#L3361-L3373

```
Literate CoffeeScript:
  ...
  extensions:
  - ".litcoffee"
  - ".coffee.md"
```

Can rmarkdown be

```
RMarkdown:
  type: prose
  color: "#198ce7"
  wrap: true
  ace_mode: markdown
  codemirror_mode: gfm
  codemirror_mime_type: text/x-gfm
  extensions:
  - ".qmd"
  - ".rmd"
  - ".q.md"    # <-- or "r.md" ?
  tm_scope: source.gfm
  language_id: 313
```

**Reply by mcanouil** [reply]
GitHub is not really an editing platform.
And changing "qmd" to "md" won't change that. It also won't preview all Quarto's features such as fenced divs and spans, tabsets, etc.

Look at https://github.com/quarto-dev/quarto-web/blob/main/index.qmd for instance, how changing to plain markdown really improves the editing experience?
On the right, what the Quarto as markdown would look like:
![image](https://github.com/user-attachments/assets/6da488a7-b9fb-45f7-ab90-a52e1a372477)

Quarto is not markdown, it is based on markdown.

If you want to edit online, I suggest to use https://github.dev/quarto-dev/quarto-web/blob/main/index.qmd
Using VSCode for web (replace `.com`, with `.dev`).


In my opinion, adding more extensions (composed extensions) won't improve things and even the opposite.

**Reply by dipterix** [reply]
> GitHub is not really an editing platform.
> And changing "qmd" to "md" won't change that. It also won't preview all Quarto's features such as fenced divs and spans, tabsets, etc.

I agree, and I'm not trying to convince you that Github edit is even comparable to quarto or trying to replace quarto. The example you showed fully demonstrates the powerful feature of quarto. That's great. 

However most of us are still doing the hybrid, and more than 50% are plain "Markdown". For those who want to have quick edits or those who don't install quarto, Github edit with markdown preview is better than coding directly with the file. (This is what I was trying to compare)

We as programmers git like breath. We care more about flexibility and powerful tools. However, still in 2024, most teams have bosses/managers who don't care about programming or markdown, they need simple-to-use **_nobrainer_** features that require zero installation of anything. Not that they don't embrace technology, they are simply too busy to care/learn/remember.

Let's see what are the steps you need to show your bosses who are coding resistant:

1. download & install quarto, or open on vscode dev
2. Git pull if running locally (you also have to make sure he/she do this every time as it's collaborative)
3. Find the correct file from git repository
4. use quarto to open and edit
5. Commit
6. Push (you have to explain to them commit is no the end of the step, often they don't remember next time)

Let's see when they use Github edit:

1. Click on `Edit this page` button provided by quarto (thanks for this feature)
2. Polish the wording, preview until they are satisfied with the paragraph
3. Save, and click OK/YES button with no brain

No need to remember anything, no need to worry about anything.

**Reply by mcanouil** [reply]
`qmd` mostly means you have computations. `qmd` extension is there to make editors aware of the code cell and provide support for them.
If you simply want to write markdown, use `md`.

Take a look at https://github.com/quarto-dev/quarto-web/blob/main/license.md, it's a plain markdown file, but it is still rendered https://quarto.org/license

**Reply by dipterix** [reply]
Programmers can set up Github action, which automatically handles the computation. 

Again it's all about collaboration in a team, not a solo project. There are people who can code but also need those who can't program to involve. 

> If you simply want to write markdown, use md.

That's what I'm doing for now, but `quarto` doesn't like it when there are computing blocks.

Can I ask what's the potential disadvantage of supporting `.q.md`, `.r.md`, or `.quarto.md`, since you said:

> In my opinion, adding more extensions (composed extensions) won't improve things and even the opposite.

It's all about balance between pros and cons. I don't see any disadvantage. All I see are advantages. If you would like a PR, I'm more than happy to propose one (just point out where I should change/what should I do).

<details>
<summary>Copilot suggestions</summary>
This is what I got from copilot. 

To make Quarto support `*.quarto.md` as a file extension (the same as `.qmd`), you should:

1. **Identify all relevant places in the codebase that handle `.qmd`**:
   - `src/resources/filters/crossref/qmd.lua`
   - `tests/smoke/manuscript/render-manuscript.test.ts`
   - `tests/docs/websites/issue-9253/_include.qmd`
   - `tests/smoke/directives/include-fixups.test.ts`
   - `dev-docs/internals-guide/render.qmd`
   
2. **Modify the code to support `*.quarto.md` extension in the identified locations**:
   - Update code to recognize `*.quarto.md` in addition to `.qmd`.
   - Add appropriate tests to ensure that `*.quarto.md` is supported in the same way as `.qmd`.

### Steps to Implement
1. **Update src/resources/filters/crossref/qmd.lua**:
   - Modify the `isQmdInput` function to recognize `*.quarto.md`.
   
2. **Update tests**:
   - Modify test cases in `tests/smoke/manuscript/render-manuscript.test.ts`, `tests/docs/websites/issue-9253/_include.qmd`, `tests/smoke/directives/include-fixups.test.ts` to include `*.quarto.md`.
   
3. **Update documentation**:
   - Reflect changes in `dev-docs/internals-guide/render.qmd`.

4. **Submit a pull request**:
   - Submit the changes and request reviews from the maintainers.
   - Ensure documentation is updated to reflect support for the new file extension.
</details>

**Reply by mcanouil** [reply]
> That's what I'm doing for now, but quarto doesn't like it when there are computing blocks.

Of course, as I said https://github.com/quarto-dev/quarto-cli/discussions/10854#discussioncomment-10820479

> Can I ask what's the potential disadvantage of supporting .q.md, .r.md, or .quarto.md, since you said:

It multiplies the number of extensions which makes it harder to identify the languages from a programmatic point of view (i.e., multiple dots extension is really horrible).
Third parties will have to add those extra extensions.
You would have to update linguist on GitHub, the "quarto" VSCode extension, RStudio IDE, Positron IDE, VIM plugin, and possibly many other places.
And in the end, you will only get very basic preview bold/italic and titles.

Note that I am not a maintainer of Quarto.

PS: I do work with non-programmers and non-GitHub users (a lot), they still managed to get edits on Quarto documents from GitHub.

**Reply by dipterix** [reply]
> Note that I am not a maintainer of Quarto.

Does that imply that you don't have decisions of whether quarto should support multi-dot extensions?

**Reply by mcanouil** [reply]
Yes.

**Reply by dipterix** [reply]
Also third parties don't need to support this extra extensions. Just need quarto to be able to compile this file format like `rmd`

> And in the end, you will only get very basic preview bold/italic and titles.

That's exactly what's needed here. Someone having programming experience will use quarto to finish the initial draft. Their bosses can always come back and edit the descriptive text.

**Reply by dipterix** [reply]
> Yes.

Then how this idea negatively affect you yourself as a user or developer. I do have reasonable motivation behind this proposal.

**Reply by mcanouil** [reply]
> Also third parties don't need to support this extra extensions.

I disagree, if a user is using `.q.md` and another user is using VIM (or any third party editors) which don't support this extension, it means this user won't be able to work on the file properly.
Making Quarto accessible especially for collaborative editing is one of the goal, one of the main feature of Quarto CLI is literate programming. You cannot break this simply to get bold/italic/titles. In the end, in the GitHub editor, all users will still have to know markdown syntax and write plain markdown.

**Reply by dipterix** [reply]
It's the thrid-party maintainer's choice on whether this should be supported. Apple does not need your agreement to stop shipping x86 CPUs. I'm only proposing supporting this extension so quarto can compile it. If the extension is not useful, it will die eventually.

I don't think it's worthwhile to tell you the benefit of using Github editor because you are not gonna use it anyway. You can't just deny it when people think something is useful/ok and you have another thing that to you is superior.

**Reply by cscheid** [reply]
I'm going to step in and call a timeout here, for everyone.

We've heard the rationale for it, and the discussion has reached the end of its utility.

I am the lead dev on Quarto, and file extension interactions are super tricky to get right because they affect multiple operating systems and existing legacy applications. We appreciate the input but at this point, we have no plan to change this.

**Reply by mcanouil** [reply]
First of all, I never said the end goal was bad or anything like that. I never said that there was a "superior" approach either.

As I said, I don't decide.
I work with non-programmers and/or non-GitHub users, they do edit Quarto documents on GitHub. That is a fact.

And may I remind you that this is a discussion board, so you might disagree with my opinion which I stated it was my opinion based on my experience, but I am still free to express it. It does not require to be Quarto main developers to have opinions based on usage and experience.

I won't comment any further as there is no longer the proper environment for that.

**Reply by dipterix** [reply]
> GitHub is not really an editing platform.
> And changing "qmd" to "md" won't change that. It also won't preview all Quarto's features such as fenced divs and spans, tabsets, etc.
> Look at [quarto-dev/quarto-web@main/index.qmd](https://github.com/quarto-dev/quarto-web/blob/main/index.qmd) for instance, how changing to plain markdown really improves the editing experience?

@mcanouil come on, what are you talking about? You were literally saying quarto editor is superior and I *agree* with that. You are such free to express it and the main dev is with your points. It's you who was promoting your points by denying the "justification" of my request.

Just like you said it's a discussion. I never said your idea is bad. Instead it's you who started with "this is a bad idea" without justifying it until I explicitly asked you "what's the disadvantage"

If you want to have a good discussion. Bring everything up front on the table and compare the pros and cons. Don't be double-standard.

**Reply by dipterix** [reply]
> I'm going to step in and call a timeout here, for everyone.
> 
> We've heard the rationale for it, and the discussion has reached the end of its utility.
> 
> I am the lead dev on Quarto, and file extension interactions are super tricky to get right because they affect multiple operating systems and existing legacy applications. We appreciate the input but at this point, we have no plan to change this.

Got it, so there is no such intention to add the support. Does `pre-render` allow you to change the file extension from `md` to `qmd`? Then I can still work on md and when it comes to compiling the project, all the `md`s are converted to `qmd`?

Just to add that in the worst-case scenarios, I can add to github action to make such conversion. Should this project not support it, it's probably fine. The devs need to have one more step to call such conversion before rendering the project locally.
